draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,76 +60,330 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+or constructed independently on clients and servers. When constructed
+independently, the mechanism used for creating the task ID must be known to
+both clients and the collector.
 
 # Leader Behavior
 
 > TODO(wangshan) Say what the Leader does when it receives a report with this
 > extension. What do Leaders do if they *don't* support this extension? Is it
 > safe to ignore it? How are the upload, aggregate, and collect flows impacted?
 
+Leader should have saved any information that is not always tied to a
+particular task, for e.g. `vdaf_verify_key`. Leader may not know the task ID of
+the current task before receiving the first upload request.
+
+## Change to update sub-protocol
+
+Upon receiving a report, leader reads the extension codepoint in
+`extension_type`. If leader does not support this extension, it SHOULD abort
+the upload protocol and alert the client with error "unrecognizedMessage".
+
+> reason for aborting: if leader ignores it, then there is no way to ensure
+> clients that the extension is used and task parameters are respected in
+> aggregators
+
+If leader supports `task-prov` extension, it should first check if the task ID
+already exists, if so leader continues to the rest of upload flow as usual.
+Note if the existing tasks's configuration is different from the one in report
+extension, HPKE decryption will fail due to mismatched AAD.
+
+If the task ID has not been seen before, leader should read and decode report's
+`extension_data` with the `TaskConfig` schema. If the decoding failed, it MUST
+abort the upload protocol and alert the client with error
+"unrecognizedMessage".
+
+If the decode succeeds, leader should create a new task using the task ID from
+the decoded report, and save task configuration with the newly created task. In
+particular, leader should deserialize `vdaf_config` corresponding to `vdaf_type`,
+and pass the relevant parameters to the VDAF initializer. At this point, the
+task provision step has completed, and leader should continue to the rest of
+upload flow.
+
+Leader MAY return error to the client if task creation failed.
+
+> OPEN ISSUE: would returning error reveal the position of the client in the
+> batch of reports received? if so is this a security or privacy threat?
+
+## Change to leader aggregate sub-protocol
+
+There is no change to the aggregate sub-protocol on leader side.
+
+## Change to collect sub-protocol
+
+During collect init, leader must perform batch validation (see batch-validation
+in {{?DAP=I-D.draft-ietf-ppm-dap-01}}) to the `CollectReq`. This requires leader
+to know the parameters associated with the DAP task. However, upon receipt of
+`CollectReq`, leader may not know the parameters if the first upload has not
+arrived yet. In this case the leader MAY respond with HTTP status code 404
+Not Found and an error of type `unrecognizedTask`. The response MAY include a
+Retry-After header field to suggest a pulling interval to the collector.
```

Comment:
Are changes actually needed? What's the goal here: to signal to the collector that the task hasn't been configured yet, but might eventually? Why is this useful?

Reply:
this is necessary because batch-validation in core protocol relies on various task parameters known at the time of receiving collectReq, the goal is to let leader have a chance to let collector know what happened, otherwise, how does leader check for e.g. min_batch_size

Reply:
Fair enough, but the following seems superfluous:

> In this case the leader MAY respond with HTTP status code 404
> Not Found and an error of type `unrecognizedTask`.

In vanilla DAP, if the task ID is not known, then the Leader is already required to abort with "unrecognizedTask". As far as I can tell, this text only changes the server behavior by allowing it to abort with 404 rather than the currently mandated 400.

> The response MAY include a
> Retry-After header field to suggest a pulling interval to the collector.

This seems fine, but a little odd. It seems like what you want to do here is distinguish between two cases:
1. A collect request is made for a task that will exist eventually (once we get the first upload request pertaining to the task)
2. A collect request is made for a task that will never exist

But how does the Leader know which of these is true, based on the information it has?

Reply:
that's a good point, we should probably aim to have collector behave the same in both cases. The core protocol today doesn't describe how leader response to a nonexistent task ID. The only `unrecognizedTask` error happens in hpke_config/ endpoint. I think the ideal solution could be the following:

- if leader receive a nonexistent task ID, it must responses 404 and `unrecognizedTask`, it may include Retry-After header.
- if collector gets 404 and `unrecognizedTask`, then it's up to collector to retry at a later time.

If this is specified in the core protocol, we can get rid of this collector section. thoughts? @cjpatton

Reply:
I agree, it seems worth spelling this out in the core protocol. Would you mind sending a PR today? If not, I will do so.

One nit: Unless there's a specific reason why 404 is desirable, let's just go with 400, which is currently used for all aborts in DAP.

Reply:
I don't have a reason to use 404 but the current core protocol actually says it explicitly: https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/blob/main/draft-ietf-ppm-dap.md#hpke-configuration-request-hpke-config

> If the aggregator does not recognize the task ID, then it responds with HTTP status code 404 Not Found and an error of type unrecognizedTask.

Reply:
In the Errors table in the core protocol, it has this:
> unrecognizedTask | An endpoint received a message with an unknown task ID.

Is this considered "spelled out" (It doesn't actually say what HTTP status code to use)? If so, then there's nothing to be done.